### ISSUE TITLE
TurboQuant KV cache silently corrupts long-context generation (immediate EOS), depends on head-of-sequence content

### ISSUE THREAD

**Issue by suntp** [Blaizzy/mlx-vlm#2162]
# Issue: TurboQuant KV cache silently corrupts long-context generation (immediate EOS), depends on head-of-sequence content

**Repo**: Blaizzy/mlx-vlm
**Version**: main @ 5c9b5f5 (0.7.0rc0), mlx 0.32.2, mlx-lm 0.31.3
**Hardware**: Apple M5 Max, 48 GB, macOS (Metal)

## Summary

With `--kv-bits 8 --kv-quant-scheme turboquant`, generation at ~92k context can produce an **empty response (first sampled token is EOS, finish_reason=stop)**. Logits are finite (no NaN/inf). The failure depends on the **content at the start of the sequence** (attention-sink region): a short marker prefix ("基准测试编号92160号，以下内容请忽略：") before a long periodic document triggers it; the same document without the marker, or with the marker moved to the end, works.

## Minimal reproduction

```python
import mlx.core as mx
from mlx_vlm import load, generate

model, processor = load("mlx-community/Qwen3.8-27B-4bit")
tok = processor.tokenizer

unit = "苹果公司发布了新一代自研芯片，中央处理器和图形处理器的性能都有明显提升，能效比也进一步优化。"
unit_ids = tok.encode(unit, add_special_tokens=False)
marker_ids = tok.encode("基准测试编号92160号，以下内容请忽略：", add_special_tokens=False)
ids = (marker_ids + unit_ids * 5000)[:92160]
prompt = tok.decode(ids) + "\n\n用一句话总结。"

res = generate(model, processor, prompt=prompt, max_tokens=8,
               temperature=0.0, kv_bits=8, kv_quant_scheme="turboquant")
print(repr(res.text), res.generation_tokens)  # -> '' 1  (expected: a summary sentence)
```

## Isolation matrix (all at ~92k tokens, temperature=0)

| case | result |
|---|---|
| marker at start + "用一句话总结。" | **EMPTY (bug)** |
| marker at start + "用一句话总结以上内容。" | works |
| no marker + either suffix | works |
| marker at end of document | works |
| same failing prompt, kv scheme `uniform` q8 | works |
| same failing prompt, f16 (no kv quant) | works |
| same failing prompt at ≤ 86k tokens | works |
| same failing prompt, `prefill_attention` forced to None (dequantize + `mx.fast.scaled_dot_product_attention` fallback) | **still EMPTY** |

## What this rules out

- Not the fused Metal attention kernels: forcing `prefill_attention` to the dequantize fallback still fails, so the corrupted data comes from the TurboQuant **encode/codec path**, not the fused score/value kernels.
- Not NaN/overflow: first-token logprobs contain no NaN; the model confidently samples EOS (top logprob ≈ -0.875).
- Not a length/indexing limit: same length works without the head marker; the failure flips with 3-token length differences combined with head content.

## Hypothesis

Head-of-sequence tokens act as attention sinks; their K/V values are typically outliers with large norms. TurboQuant currently quantizes them like any other token (no sink-token protection / precision preservation), so at long contexts the sink-region quantization error corrupts the global attention distribution enough for EOS to win argmax. llama.cpp/vLLM-style KV quantization keeps sink/early tokens in higher precision for exactly this reason.

Suggested directions:
1. Keep the first N tokens' K/V unquantized in `TurboQuantKVCache` (sink protection), or
2. Investigate outlier-norm handling in `_TurboQuantProdCodec` key encoding.

## Notes

- Reproducible 100% deterministically (temperature=0, fixed TurboQuant seed).
- peak memory at failure: 22.6 GiB (not memory-pressure related; reproduced with 60%+ system RAM free).

**Comment by suntp** [comment]
**Correction / closing: this is not a TurboQuant defect — the failing input is a knife-edge case that any 8-bit KV quantization flips.**

Follow-up investigation after filing turned up evidence that my original framing was wrong:

1. **uniform q8 fails identically** on the same prompt (empty output, first token `<|im_end|>`) — with stock code, with `--kv-group-size 32`, and regardless of the fused/dequantize attention path. Not specific to the TurboQuant codec.
2. **The f16 baseline is itself on a knife's edge here**: top-3 first-token logprobs are `"\n\n"` -1.125, `<think>` -1.125, `<|im_end|>` -1.25. EOS trails the correct answer by only ~0.125 natively, so a ~0.25-logit shift — well within normal q8 KV noise at 92k context — flips argmax to EOS.
3. The apparent content/length dependence is just the shape of that decision boundary.
4. Cross-engine note: llama.cpp with q8_0 KV (`--cache-type-k q8_0 --cache-type-v q8_0`) answers this prompt correctly — but that comparison is confounded by different weight quantization (GGUF Q4_K_M vs MLX 4-bit), so it is not evidence of an MLX-side defect.

Nothing is broken in TurboQuant. 8-bit KV quantization perturbs logits enough to flip a borderline synthetic prompt (92k tokens of near-identical repeated text + marker prefix is a worst case for attention-score noise). Closing as not-a-bug; apologies for the noise.

The real, reproducible issues found during the same investigation (quantized-KV peak-memory/OOM at long context, and multi-turn corruption from post-hoc `to_quantized()` conversion) are tracked in #2163 with a working patch.
